src/Compilers/CSharp/Portable/Binder/Semantics/Operators/BinaryOperatorOverloadResolution.cs

```diff
@@ -825,25 +834,95 @@ private static LiftingResult UserDefinedBinaryOperatorCanBeLifted(TypeSymbol lef
                     {
                         continue;
                     }
+
                     var better = BetterOperator(candidates[i].Signature, candidates[j].Signature, left, right, ref useSiteDiagnostics);
-                    if (better == BetterResult.Left)
+                    if (better == BetterResult.Left && candidates[j].Kind != OperatorAnalysisResultKind.Worse)
                     {
                         candidates[j] = candidates[j].Worse();
                     }
-                    else if (better == BetterResult.Right)
+                    else if (better == BetterResult.Right && candidates[i].Kind != OperatorAnalysisResultKind.Worse)
                     {
                         candidates[i] = candidates[i].Worse();
                     }
                 }
             }
         }
 
-        private BetterResult BetterOperator(BinaryOperatorSignature op1, BinaryOperatorSignature op2, BoundExpression left, BoundExpression right, ref HashSet<DiagnosticInfo> useSiteDiagnostics)
+        private bool EliminateWorseCandidates_FastPath(
```

Comment:
>EliminateWorseCandidates_FastPath [](start = 21, length = 33)

I think the "fast" path shouldn't be changing kind of candidates unless it returns true. Basically it should first find the best candidate and, only then, mark all other applicable candidates as Worse. Marking candidates as Worse in case when we cannot find a single best can negatively affect result of ```EliminateWorseCandidates_SlowPath``` because it will never compare a candidate marked Worse with candidates preceding it, thus won't perform the same analysis as performed today without the optimization.